internal/controller/ec2.aws/securitygroup_controller.go

```diff
@@ -161,86 +179,101 @@ func (r *SecurityGroupReconciliation) retrieveInfraRefInfo(ctx context.Context)
+				Namespace: infrastructureRef.Namespace,
+			}
+			if err := r.Client.Get(ctx, key, &kmp); err != nil {
+				retrieveErr = err
+				continue
+			}
 
-		kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
-		key = client.ObjectKey{
-			Name:      kmp.Spec.ClusterName,
-			Namespace: kmp.ObjectMeta.Namespace,
-		}
-		if err := r.Client.Get(ctx, key, kcp); err != nil {
-			return err
-		}
+			kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
+			key = client.ObjectKey{
+				Name:      kmp.Spec.ClusterName,
+				Namespace: kmp.ObjectMeta.Namespace,
+			}
+			if err := r.Client.Get(ctx, key, kcp); err != nil {
+				retrieveErr = err
+				continue
+			}
 
-		region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
-		if err != nil {
-			return fmt.Errorf("error retrieving region: %w", err)
-		}
+			region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
+			if err != nil {
+				retrieveErr = fmt.Errorf("error retrieving region: %w", err)
+				continue
+			}
 
-		providerConfigName, awsCfg, err := kopsutils.RetrieveAWSCredentialsFromKCP(ctx, r.Client, region, kcp)
-		if err != nil {
-			return err
-		}
+			providerConfigName, awsCfg, err := kopsutils.RetrieveAWSCredentialsFromKCP(ctx, r.Client, region, kcp)
+			if err != nil {
+				retrieveErr = err
+				continue
+			}
 
-		r.ec2Client = r.NewEC2ClientFactory(*awsCfg)
-		r.asgClient = r.NewAutoScalingClientFactory(*awsCfg)
+			r.ec2Client = r.NewEC2ClientFactory(*awsCfg)
+			r.asgClient = r.NewAutoScalingClientFactory(*awsCfg)
 
-		vpcId, err := ec2.GetVPCIdWithCIDRAndClusterName(ctx, r.ec2Client, kcp.Name, kcp.Spec.KopsClusterSpec.Networking.NetworkCIDR)
-		if err != nil {
-			return fmt.Errorf("error retrieving vpcID: %w", err)
-		}
+			vpcId, err := ec2.GetVPCIdWithCIDRAndClusterName(ctx, r.ec2Client, kcp.Name, kcp.Spec.KopsClusterSpec.Networking.NetworkCIDR)
+			if err != nil {
+				retrieveErr = fmt.Errorf("error retrieving vpcID: %w", err)
+				continue
+			}
 
-		r.providerConfigName = providerConfigName
-		r.region = region
-		r.vpcId = vpcId
+			r.providerConfigName = providerConfigName
+			r.region = region
+			r.vpcId = vpcId
 
-		return nil
-	case "KopsControlPlane":
-		kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
-		key := client.ObjectKey{
-			Name:      r.sg.Spec.InfrastructureRef[0].Name,
-			Namespace: r.sg.Spec.InfrastructureRef[0].Namespace,
-		}
-		if err := r.Client.Get(ctx, key, kcp); err != nil {
-			return err
-		}
+			return nil
+		case "KopsControlPlane":
+			kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
+			key := client.ObjectKey{
+				Name:      infrastructureRef.Name,
+				Namespace: infrastructureRef.Namespace,
+			}
+			if err := r.Client.Get(ctx, key, kcp); err != nil {
+				retrieveErr = err
+				continue
+			}
 
-		region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
-		if err != nil {
-			return fmt.Errorf("error retrieving region: %w", err)
-		}
+			region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
```

Comment:
this part is quite similar to the one above, the only difference between them is the retrieval of the kmp. WDYT about encapsulating what is similar between them?

Reply:
makes sense, this could be a private function

Reply:
fixed in the latest changes, but only extracted the repeated kcp part that was repeated